### PR TITLE
StringHelper::pregExplode implemented

### PR DESCRIPTION
**Pull request by SilverFire**
Propose to add pregExplode method to the `StringHelper`.
The difference with `StringHelper::explode` is that this one is for splitting using regular expression.

### CODE REVIEW

File: framework/helpers/BaseStringHelper.php [hunk 1]
```diff
@@ -268,4 +268,37 @@ public static function explode($string, $delimiter = ',', $trim = true, $skipEmp
...
+            } elseif (!is_callable($trim)) {
+                $trim = function ($v) use ($trim) {
+                    return trim($v, $trim);
+                };
+            }
+            $result = array_map($trim, $result);
+        }
+        if ($skipEmpty) {
+            // Wrapped with array_values to make array keys sequential after empty values removing
+            $result = array_values(array_filter($result));
```
Comment (kidol) [inline framework/helpers/BaseStringHelper.php]: array_filter will also remove string `0`, I doubt that's expected? https://3v4l.org/dXpPr

Comment (SilverFire) [inline framework/helpers/BaseStringHelper.php]: Aw, snap! It really does! Thank you for your attention!

Comment (SilverFire) [inline framework/helpers/BaseStringHelper.php]: Fixed. Thank you very much

File: framework/helpers/BaseStringHelper.php [hunk 2]
```diff
@@ -270,4 +270,39 @@ public static function explode($string, $delimiter = ',', $trim = true, $skipEmp
...
+     * @return array
+     * @since 2.0.7
+     */
+    public static function pregExplode($string, $pattern = '/[\s,;]+/', $trim = true, $skipEmpty = false)
+    {
+        $result = preg_split($pattern, $string);
+        if ($trim) {
+            if ($trim === true) {
+                $trim = 'trim';
+            } elseif (!is_callable($trim)) {
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php]: It's not necessary. There could be no situation where `trim()` doesn't exist.

File: framework/helpers/BaseStringHelper.php [hunk 3]
```diff
@@ -270,4 +270,39 @@ public static function explode($string, $delimiter = ',', $trim = true, $skipEmp
...
+                $trim = function ($v) use ($trim) {
+                    return trim($v, $trim);
+                };
+            }
+            $result = array_map($trim, $result);
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php]: Condition at line 290 could be moved here.

File: framework/helpers/BaseStringHelper.php [hunk 4]
```diff
@@ -270,4 +270,39 @@ public static function explode($string, $delimiter = ',', $trim = true, $skipEmp
...
+        }
+        if ($skipEmpty) {
+            // Wrapped with array_values to make array keys sequential after empty values removing
+            $result = array_values(array_filter($result, function ($value) {
```
Comment (samdark) [inline framework/helpers/BaseStringHelper.php]: There's `PREG_SPLIT_NO_EMPTY` for that.

**Comment by samdark** [comment]
After all adjustments isn't becoming just `preg_split` with defaults?

**Comment by SilverFire** [comment]
You're right.

**Pull request closed without merging**